hare3/hare.go

```diff
@@ -0,0 +1,571 @@
+package hare3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/datastore"
+	"github.com/spacemeshos/go-spacemesh/metrics"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/ballots"
+	"github.com/spacemeshos/go-spacemesh/sql/beacons"
+	"github.com/spacemeshos/go-spacemesh/sql/identities"
+	"github.com/spacemeshos/go-spacemesh/sql/proposals"
+	"github.com/spacemeshos/go-spacemesh/system"
+	"github.com/spacemeshos/go-spacemesh/timesync"
+)
+
+type Config struct {
+	Enable          bool          `mapstructure:"enable"`
+	EnableAfter     types.LayerID `mapstructure:"enable-layer"`
+	Committee       uint16        `mapstructure:"committee"`
+	Leaders         uint16        `mapstructure:"leaders"`
+	IterationsLimit uint8         `mapstructure:"iterations-limit"`
+	PreroundDelay   time.Duration `mapstructure:"preround-delay"`
+	RoundDuration   time.Duration `mapstructure:"round-duration"`
+	ProtocolName    string
+}
+
+func (cfg *Config) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
+	encoder.AddBool("enabled", cfg.Enable)
+	encoder.AddUint32("after", cfg.EnableAfter.Uint32())
+	encoder.AddUint16("committee", cfg.Committee)
+	encoder.AddUint16("leaders", cfg.Leaders)
+	encoder.AddUint8("iterations limit", cfg.IterationsLimit)
+	encoder.AddDuration("preround delay", cfg.PreroundDelay)
+	encoder.AddDuration("round duration", cfg.RoundDuration)
+	encoder.AddString("p2p protocol", cfg.ProtocolName)
+	return nil
+}
+
+// roundStart returns expected time for iter/round relative to
+// layer start.
+func (cfg *Config) roundStart(round IterRound) time.Duration {
+	duration := cfg.PreroundDelay
+	if round.Round == preround {
+		return duration
+	}
+	delay := time.Duration(round.Delay(IterRound{Round: preround})) - 1
+	return duration + cfg.RoundDuration*delay
+}
+
+func DefaultConfig() Config {
+	return Config{
+		Committee:       800,
+		Leaders:         10,
+		IterationsLimit: 40,
```

Comment:
this limit is based on iterations limit, why would i add another number?

Reply:
my assumption was that 40 iterations is a security parameter. that's why i didn't think it's good to mix the two params.

Reply:
but if there will be additional parameter for max concurrent iterations, then it will have to overwrite iterations limit right? so i would rather change 40 iterations. for example i definitely don't want to start with 40 on mainnet